### PR TITLE
Backport 2.3 changes

### PR DESCRIPTION
**Pull request by greg0ire**
~~I'm just using travis to test this branch.~~ Can and should be merged when the build is passing, in fact.

### CODE REVIEW

File: Resources/doc/reference/security.rst [hunk 1]
```diff
@@ -31,7 +31,7 @@ The install process is available on the dedicated
 Security handlers
 -----------------
 
-The security part is managed by a ``SecurityHandler``, the bundle comes with 3 handlers
+The security part is managed by a ``SecurityHandler``, the bundle comes with 3 handlers:
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 2]
```diff
@@ -104,7 +104,7 @@ Configuration
 
 First, activate the role security handler as described above.
 
-Each time an user tries to do an action in the admin, Sonata checks if he is
+Each time a user tries to do an action in the admin, Sonata checks if he is
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 3]
```diff
@@ -129,7 +129,7 @@ The role name will be based on the name of your admin service. For instance, ``a
 
     If your admin service is named like ``my.blog.admin.foo_bar`` (note the underscore ``_``) it will become: ``ROLE_MY_BLOG_ADMIN_FOO_BAR_{ACTION}``
 
-So our ``security.yml`` file may look to something like this:
+So our ``security.yml`` file may look something like this:
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 4]
```diff
@@ -310,7 +310,7 @@ The following configuration for the SonataUserBundle defines:
 - the login form for authentication
 - the access control: resources with related required roles, the important
   part is the admin configuration
-- the ``acl`` option to enable the ACL.
+- the ``acl`` option to enable the ACL
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 5]
```diff
@@ -459,10 +459,10 @@ Owner:
 Vocabulary used for Access Control Lists:
 
 - **Role:** a user role;
-- **ACL:** a list of access rules, the Admin uses 2 types:
+- **ACL:** a list of access rules, the Admin uses 2 types;
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 6]
```diff
@@ -459,10 +459,10 @@ Owner:
...
 - **Admin ACL:** created from the Security information of the Admin class
   for  each admin and shares the Access Control Entries that specify what
-  the  user can do (permissions) with the admin
+  the  user can do (permissions) with the admin;
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 7]
```diff
@@ -497,9 +497,9 @@ In the application the security context is asked if access is granted for a role
 or a permission (``admin.isGranted``):
 
 - **Token:** a token identifies a user between requests;
-- **Voter:** sort of judge that returns if access is granted of denied, if the
+- **Voter:** sort of judge that returns whether access is granted or denied, if the
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Resources/doc/reference/security.rst [hunk 8]
```diff
@@ -497,9 +497,9 @@ In the application the security context is asked if access is granted for a role
...
   voter should not vote for a case, it returns abstain;
-- **AccessDecisionManager:** decides if access is granted or denied according
+- **AccessDecisionManager:** decides whether access is granted or denied according
```
Comment (greg0ire) [inline Resources/doc/reference/security.rst]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 9]
```diff
@@ -116,8 +116,6 @@ public function setUp()
         $this->environment->addExtension(new RoutingExtension($urlGenerator));
         $this->environment->addExtension(new \Twig_Extensions_Extension_Text());
 
-        $this->twigExtension->initRuntime($this->environment);
-
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 10]
```diff
@@ -247,7 +245,7 @@ public function testRenderListElement($expected, $type, $value, array $options)
                 }
             }));
 
-        $this->assertSame($expected, trim(preg_replace('/\s+/', ' ', $this->twigExtension->renderListElement($this->object, $this->fieldDescription))));
+        $this->assertSame($expected, trim(preg_replace('/\s+/', ' ', $this->twigExtension->renderListElement($this->environment, $this->object, $this->fieldDescription))));
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 11]
```diff
@@ -407,7 +405,7 @@ public function testRenderListElementNonExistentTemplate()
             ->method('warning')
             ->with(($this->stringStartsWith('An error occured trying to load the template "SonataAdminBundle:CRUD:list_nonexistent_template.html.twig" for the field "Foo_name", the default template "SonataAdminBundle:CRUD:base_list_field.html.twig" was used instead: "Unable to find template "list_nonexistent_template.html.twig')));
 
-        $this->twigExtension->renderListElement($this->object, $this->fieldDescription);
+        $this->twigExtension->renderListElement($this->environment, $this->object, $this->fieldDescription);
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 12]
```diff
@@ -425,7 +423,7 @@ public function testRenderListElementErrorLoadingTemplate()
             ->method('getTemplate')
             ->will($this->returnValue('SonataAdminBundle:CRUD:list_nonexistent_template.html.twig'));
 
-        $this->twigExtension->renderListElement($this->object, $this->fieldDescription);
+        $this->twigExtension->renderListElement($this->environment, $this->object, $this->fieldDescription);
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 13]
```diff
@@ -486,7 +484,7 @@ public function testRenderViewElement($expected, $type, $value, array $options)
                 }
             }));
 
-        $this->assertSame($expected, trim(preg_replace('/\s+/', ' ', $this->twigExtension->renderViewElement($this->fieldDescription, $this->object))));
+        $this->assertSame($expected, trim(preg_replace('/\s+/', ' ', $this->twigExtension->renderViewElement($this->environment, $this->fieldDescription, $this->object))));
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 14]
```diff
@@ -702,11 +700,11 @@ public function testOutput()
         $template = $this->environment->loadTemplate('SonataAdminBundle:CRUD:base_list_field.html.twig');
 
         $this->assertSame('<td class="sonata-ba-list-field sonata-ba-list-field-" objectId="12345"> foo </td>',
-                trim(preg_replace('/\s+/', ' ', $this->twigExtension->output($this->fieldDescription, $template, $parameters))));
+                trim(preg_replace('/\s+/', ' ', $this->twigExtension->output($this->fieldDescription, $template, $parameters, $this->environment))));
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 15]
```diff
@@ -702,11 +700,11 @@ public function testOutput()
...
 
         $this->environment->enableDebug();
         $this->assertSame('<!-- START fieldName: fd_name template: SonataAdminBundle:CRUD:base_list_field.html.twig compiled template: SonataAdminBundle:CRUD:base_list_field.html.twig --> <td class="sonata-ba-list-field sonata-ba-list-field-" objectId="12345"> foo </td> <!-- END - fieldName: fd_name -->',
-                trim(preg_replace('/\s+/', ' ', $this->twigExtension->output($this->fieldDescription, $template, $parameters))));
+                trim(preg_replace('/\s+/', ' ', $this->twigExtension->output($this->fieldDescription, $template, $parameters, $this->environment))));
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 16]
```diff
@@ -26,11 +26,6 @@
 class SonataAdminExtension extends \Twig_Extension implements \Twig_Extension_InitRuntimeInterface
 {
     /**
-     * @var \Twig_Environment
-     */
-    protected $environment;
-
-    /**
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 17]
```diff
@@ -53,19 +48,11 @@ public function __construct(Pool $pool, LoggerInterface $logger = null)
...
      * {@inheritdoc}
      */
-    public function initRuntime(\Twig_Environment $environment)
-    {
-        $this->environment = $environment;
-    }
-
-    /**
-     * {@inheritdoc}
-     */
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 18]
```diff
@@ -53,19 +48,11 @@ public function __construct(Pool $pool, LoggerInterface $logger = null)
...
     public function getFilters()
     {
         return array(
-            new \Twig_SimpleFilter('render_list_element', array($this, 'renderListElement'), array('is_safe' => array('html'))),
-            new \Twig_SimpleFilter('render_view_element', array($this, 'renderViewElement'), array('is_safe' => array('html'))),
+            new \Twig_SimpleFilter('render_list_element', array($this, 'renderListElement'), array('is_safe' => array('html'), 'needs_environment' => true)),
+            new \Twig_SimpleFilter('render_view_element', array($this, 'renderViewElement'), array('is_safe' => array('html'), 'needs_environment' => true)),
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 19]
```diff
@@ -89,14 +76,14 @@ public function getName()
      *
      * @return \Twig_Template
      */
-    protected function getTemplate(FieldDescriptionInterface $fieldDescription, $defaultTemplate)
+    protected function getTemplate(FieldDescriptionInterface $fieldDescription, $defaultTemplate, Twig_Environment $environment)
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 20]
```diff
@@ -89,14 +76,14 @@ public function getName()
...
     {
         $templateName = $fieldDescription->getTemplate() ?: $defaultTemplate;
 
         try {
-            $template = $this->environment->loadTemplate($templateName);
+            $template = $environment->loadTemplate($templateName);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 21]
```diff
@@ -89,14 +76,14 @@ public function getName()
...
         } catch (\Twig_Error_Loader $e) {
-            $template = $this->environment->loadTemplate($defaultTemplate);
+            $template = $environment->loadTemplate($defaultTemplate);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 22]
```diff
@@ -115,16 +102,16 @@ protected function getTemplate(FieldDescriptionInterface $fieldDescription, $def
      *
      * @return string
      */
-    public function renderListElement($object, FieldDescriptionInterface $fieldDescription, $params = array())
+    public function renderListElement(Twig_Environment $environment, $object, FieldDescriptionInterface $fieldDescription, $params = array())
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 23]
```diff
@@ -115,16 +102,16 @@ protected function getTemplate(FieldDescriptionInterface $fieldDescription, $def
...
     {
-        $template = $this->getTemplate($fieldDescription, $fieldDescription->getAdmin()->getTemplate('base_list_field'));
+        $template = $this->getTemplate($fieldDescription, $fieldDescription->getAdmin()->getTemplate('base_list_field'), $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 24]
```diff
@@ -115,16 +102,16 @@ protected function getTemplate(FieldDescriptionInterface $fieldDescription, $def
...
 
         return $this->output($fieldDescription, $template, array_merge($params, array(
             'admin'             => $fieldDescription->getAdmin(),
             'object'            => $object,
             'value'             => $this->getValueFromFieldDescription($object, $fieldDescription),
             'field_description' => $fieldDescription,
-        )));
+        )), $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 25]
```diff
@@ -134,11 +121,11 @@ public function renderListElement($object, FieldDescriptionInterface $fieldDescr
      *
      * @return string
      */
-    public function output(FieldDescriptionInterface $fieldDescription, \Twig_Template $template, array $parameters = array())
+    public function output(FieldDescriptionInterface $fieldDescription, \Twig_Template $template, array $parameters = array(), Twig_Environment $environment)
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 26]
```diff
@@ -134,11 +121,11 @@ public function renderListElement($object, FieldDescriptionInterface $fieldDescr
...
     {
         $content = $template->render($parameters);
 
-        if ($this->environment->isDebug()) {
+        if ($environment->isDebug()) {
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 27]
```diff
@@ -189,9 +176,9 @@ public function getValueFromFieldDescription($object, FieldDescriptionInterface
      *
      * @return string
      */
-    public function renderViewElement(FieldDescriptionInterface $fieldDescription, $object)
+    public function renderViewElement(Twig_Environment $environment, FieldDescriptionInterface $fieldDescription, $object)
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 28]
```diff
@@ -189,9 +176,9 @@ public function getValueFromFieldDescription($object, FieldDescriptionInterface
...
     {
-        $template = $this->getTemplate($fieldDescription, 'SonataAdminBundle:CRUD:base_show_field.html.twig');
+        $template = $this->getTemplate($fieldDescription, 'SonataAdminBundle:CRUD:base_show_field.html.twig', $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 29]
```diff
@@ -204,7 +191,7 @@ public function renderViewElement(FieldDescriptionInterface $fieldDescription, $
             'object'            => $object,
             'value'             => $value,
             'admin'             => $fieldDescription->getAdmin(),
-        ));
+        ), $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 30]
```diff
@@ -216,9 +203,9 @@ public function renderViewElement(FieldDescriptionInterface $fieldDescription, $
      *
      * @return string
      */
-    public function renderViewElementCompare(FieldDescriptionInterface $fieldDescription, $baseObject, $compareObject)
+    public function renderViewElementCompare(Twig_Environment $environment, FieldDescriptionInterface $fieldDescription, $baseObject, $compareObject)
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 31]
```diff
@@ -216,9 +203,9 @@ public function renderViewElement(FieldDescriptionInterface $fieldDescription, $
...
     {
-        $template = $this->getTemplate($fieldDescription, 'SonataAdminBundle:CRUD:base_show_field.html.twig');
+        $template = $this->getTemplate($fieldDescription, 'SonataAdminBundle:CRUD:base_show_field.html.twig', $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Twig/Extension/SonataAdminExtension.php [hunk 32]
```diff
@@ -253,7 +240,7 @@ public function renderViewElementCompare(FieldDescriptionInterface $fieldDescrip
             'value_compare'     => $compareValue,
             'is_diff'           => $isDiff,
             'admin'             => $fieldDescription->getAdmin(),
-        ));
+        ), $environment);
```
Comment (greg0ire) [inline Twig/Extension/SonataAdminExtension.php]: :+1:

File: Admin/Admin.php [hunk 33]
```diff
@@ -57,7 +57,7 @@
     const CONTEXT_MENU       = 'menu';
     const CONTEXT_DASHBOARD  = 'dashboard';
 
-    const CLASS_REGEX        = '@(?:([A-Za-z0-9]*)\\\)?(Bundle\\\)?([A-Za-z0-9]+)Bundle\\\(Entity|Document|Model|PHPCR|CouchDocument|Phpcr|Doctrine\\\Orm|Doctrine\\\Phpcr|Doctrine\\\MongoDB|Doctrine\\\CouchDB)\\\(.*)@';
+    const CLASS_REGEX        = '@(?:([A-Za-z0-9]*)\\\)?(Bundle\\\)?([A-Za-z0-9]+?)(?:Bundle)?\\\(Entity|Document|Model|PHPCR|CouchDocument|Phpcr|Doctrine\\\Orm|Doctrine\\\Phpcr|Doctrine\\\MongoDB|Doctrine\\\CouchDB)\\\(.*)@';
```
Comment (greg0ire) [inline Admin/Admin.php]: Introduced in 61612242665125cf88f117d88ce59d6ab698d4dd

**Comment by greg0ire** [comment]
@wodka : can you have a look at the build and tell me how to repair it? The current error has to do with your changes…

File: Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php [hunk 34]
```diff
@@ -55,6 +55,8 @@ public function testEmbeddedAdmin()
             $meta->tags['sonata.admin'][0],
             array(
                 'manager_type'      => 'orm',
+                'group'             => 'Admin',
+                'label'             => 'Tests\Fixtures\Foo',
```
Comment (greg0ire) [inline Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php]: @wodka : can you validate this ?

File: composer.json [hunk 35]
```diff
@@ -40,6 +40,10 @@
         "doctrine/inflector": "~1.0",
         "knplabs/knp-menu-bundle": "^2.1.1"
     },
+    "conflict": {
+        "jms/di-extra-bundle": "<1.7.0",
+        "symfony/security": ">2.8.999"
+    },
```
Comment (greg0ire) [inline composer.json]: :-1:

File: Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php (hunk 34 above)
Comment (wodka) [inline Tests/DependencyInjection/Compiler/AnnotationCompilerPassTest.php]: this should be ok (based on the regex change of https://github.com/sonata-project/SonataAdminBundle/pull/3687#discussion_r57702074)

this is just a fallback for the admin name

File: Admin/Admin.php (hunk 33 above)
Comment (greg0ire) [inline Admin/Admin.php]: @wodka this is the change that is was not compatible with your tests.

**Comment by OskarStark** [comment]
:+1: thank you guys!

File: .travis.yml [hunk 36]
```diff
@@ -7,7 +7,7 @@ sudo: false
 
 cache:
   directories:
-    - $HOME/.composer/cache
+    - $HOME/.composer/cache/files
```
Comment (greg0ire) [inline .travis.yml]: @Koc do you validate this backport ?

Comment (Koc) [inline .travis.yml]: Yes, it is better cache this folder

https://github.com/FriendsOfSymfony/FOSRestBundle/blob/2.0/.travis.yml#L13
https://github.com/liip/LiipImagineBundle/blob/master/.travis.yml

Comment (greg0ire) [inline .travis.yml]: Thanks for the quick feedback!

**Comment by Koc** [comment]
@greg0ire nice work, it is good backport bugfixes to master from 2.3

**Comment by greg0ire** [comment]
@Koc thanks! It is quite hard though if we don't do it regularly, and most of the time are backported from master instead of the other way around. I think we should be more careful about that in PRs.

**Pull request merged**